finance/mip_raw_elt_batch.py

```diff
@@ -0,0 +1,285 @@
+LOG.setLevel(logging.DEBUG)
+
+_mips_url_login = "https://login.mip.com/api/v1/sso/mipadv/login"
+# https://documentation.mip.com/#/swagger/recursiveTransaction/transactions/recursiveTransactions
+_ledger_api = "https://api.mip.com/api/v1/recursiveTransactions/GeneralLedger/posted"
+# https://documentation.mip.com/#/swagger/maintenance/generalLedger/chartOfAccounts
+_chart_of_accounts_api = "https://api.mip.com/api/v1/maintain/ChartOfAccounts"
+_mips_url_logout = "https://api.mip.com/api/security/logout"
+
+
+@backoff.on_exception(backoff.expo, (RequestError, RequestException), max_time=11)
+def _request_login(creds):
+    """
+    Wrap login request with backoff decorator, using exponential backoff
+    and running for at most 11 seconds. With a connection timeout of 4
+    seconds, this allows two attempts.
+    """
+    timeout = 4
+    LOG.info("Logging in to upstream API")
+
+    login_response = requests.post(
+        _mips_url_login,
+        json=creds,
+        timeout=timeout,
+    )
+    login_response.raise_for_status()
+    token = login_response.json()["AccessToken"]
+    return token
+
+
+@backoff.on_exception(backoff.fibo, (RequestError, RequestException), max_time=28)
+def _request_logout(access_token):
+    """
+    Wrap logout request with backoff decorator, using fibonacci backoff
+    and running for at most 28 seconds. With a connection timeout of 6
+    seconds, this allows three attempts.
+
+    Prioritize spending time logging out over the other requests because
+    failing to log out after successfully logging in will lock us out of
+    the API; but CloudFront will only wait a maximum of 60 seconds for a
+    response from this lambda.
+    """
+    timeout = 6
+    LOG.info("Logging out of upstream API")
+
+    requests.post(
+        _mips_url_logout,
+        headers={"Authorization-Token": access_token},
+        timeout=timeout,
+    )
+
+
+def get_ledgers(
+    access_token: str,
+    max_session_posted_date: str = "2025-01-24 23:48:49.000",
+    page_number: int = 0,
+    page_size: int = 20,
+):
+    """Get ledgers from the MIP API
+
+    Args:
+        access_token (str): MIP access token
+        max_session_posted_date (str, optional): Max session posted date, used to filter for transactions after a specific date. Defaults to "2025-01-24 23:48:49.000".
+        page_number (int, optional): _description_. Defaults to 0.
+        page_size (int, optional): _description_. Defaults to 20.
+
+    Returns:
+        _type_: _description_
+    """
+    all_ledgers = []
+    initial_ledgers = requests.get(
+        _ledger_api,
+        headers={
+            "Authorization-Token": access_token,
+        },
+        params={
+            "page[number]": str(page_number),
+            "page[size]": str(page_size),
+            "filter[sessionPostedDate][gt]": max_session_posted_date,
+        },
+    )
+    total_records = initial_ledgers.json()["totalRecords"]
+    while len(all_ledgers) < total_records:
```

Comment:
team braphilianna: If you ever want this loop to exit, let's hope that

1. The `data` field is never empty.
2. Nobody passes too large of a `page_number` to this function.